### ISSUE TITLE
Add integration test

### ISSUE THREAD

**Issue by rodrigosiqueira** [kworkflow/kworkflow#262]
Currently, we only have unit tests. Let's start to investigate how to add an integration test; some random ideas:

1. Create an integration test setup; maybe, we can use virtual machines for that;
2. Start with a simple feature like `kw b`;
3. Find a way to validate if something happened as expected;
4. Provide a report;
5. Can we integrate it with Travis or similar?

**Comment by alan-barzilay** [comment]
Maybe we could use docker containers for the tests, I heard that they are easier to use with Travis than virtual machines

**Comment by alan-barzilay** [comment]
Scratch that, docker uses the same kernel as the host system so it will probably have limited use for us. Bu since 2019 Vagrant seems to be supported in travis https://stackoverflow.com/questions/37307114/launching-a-vagrant-vm-inside-travis-ci

**Comment by rodrigosiqueira** [comment]
Nice,
Maybe we can also use another type of CI that facilitates running an integration CI. It would be nice if we have a pipeline for these things.

**Comment by rodrigosiqueira** [comment]
Hi @aquilamacedo,

After passing through your PR, you gave me a lot of ideas, and I'll try to summarize how we can approach this problem.

# Setup phase

Integration tests will require a lot of setup to work as we want. With this idea in mind, I suggested creating a prepare function in the run_integration test. You could add a `--prepare` option to the integration script, but you can also call it by default in case the user just calls `run_integration_test.sh`. This function should do the following:

1. Setup any folder used by the integration test;
2. Download any container or VM necessary for the test;
3. Conduct any setup necessary in the container or VM.

# Run phase

We should only try to run the integration tests if the setup criteria are fulfilled. For the run phase, we will need to create tools to work together with shunit. For example:

1. We will need a library to get data from the container/VM.
2. We will probably need ways to kill a container/VM in case of failure.

# Test method

We need an easy way to write tests, and we can have a good blend between in-house tools and shunit. Instead of trying to create this ideal scenario, let's only focus on test kw build. We could have something like this pseudo-code:

```sh
#!/bin/bash

. ../src/lib/kw_include.sh --source-only
include '../src/lib/kwio.sh'
include './utils.sh'

os='debian'

function setUp()
{
  start_clean_container
  configure_kernel_tree_in_the_container
  install_current_kw_version_in_the_container
  cd kernel_inside_container
  setup_kw_for_build_test
}

function tearDown()
{
  stop_container
}

function test_simple_build()
{
   local output

   run_command_in_the_container('cd path/to/kernel && kw build')
 
   # Check if the kernel build happened
   check_return_code_error
   check_if_the_kernel_image_was_create
   check_statistic
}

invoke_shunit

```

# Manager our own VM images and containers

We should not rely on external images, but we also need to find a scalable way for us to maintain them.

# Simple concept for the next PR

For the next PR, let's keep things under control and only add one integration test: a simple case of building a kernel. It would help if you had something like this:

1. Commit 1: Create the setup part.
2. Commit 2: Create the first test.
3. Commit 3: Ensure you have a good documentation.

After you finish this part, we should start another PR in parallel to create the required scripts to maintain our VMs/containers.

**Comment by fernandohjunqueira** [comment]
We (me, @uwla, @gustavofukunaga) are working on implementing integration tests with Podman. This is our PR #953.

**Comment by rodrigosiqueira** [comment]
Thanks @uwla and @fernandohjunqueira . We now have integration test :)
